components/brave_ads/core/internal/ads/serving/choose/predict_ad_embeddings.h

```diff
@@ -0,0 +1,58 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_ADS_CORE_INTERNAL_ADS_SERVING_CHOOSE_PREDICT_AD_EMBEDDINGS_H_
+#define BRAVE_COMPONENTS_BRAVE_ADS_CORE_INTERNAL_ADS_SERVING_CHOOSE_PREDICT_AD_EMBEDDINGS_H_
+
+#include <vector>
+
+#include "absl/types/optional.h"
+#include "base/check_op.h"
+#include "base/rand_util.h"
+#include "brave/components/brave_ads/core/internal/ads/serving/choose/eligible_ads_predictor_util.h"
+#include "brave/components/brave_ads/core/internal/ads/serving/choose/sample_ads.h"
+#include "brave/components/brave_ads/core/internal/ads/serving/eligible_ads/pacing/pacing.h"
+#include "brave/components/brave_ads/core/internal/ads/serving/targeting/user_model_info.h"
+#include "brave/components/brave_ads/core/internal/common/numbers/number_util.h"
+#include "brave/components/brave_ads/core/internal/processors/contextual/text_embedding/text_embedding_html_event_info.h"
+
+namespace ads {
+
+template <typename T>
+absl::optional<T> MaybePredictAdUsingEmbeddings(
+    const targeting::UserModelInfo& user_model,
+    const std::vector<T>& creative_ads) {
+  DCHECK(!creative_ads.empty());
+
+  const std::vector<T> paced_creative_ads = PaceCreativeAds(creative_ads);
+
+  if (paced_creative_ads.empty()) {
+    return absl::nullopt;
+  }
+
+  const std::vector<int> votes_registry =
+      ComputeVoteRegistry(paced_creative_ads, user_model.embeddings_history);
+
+  DCHECK_EQ(votes_registry.size(), paced_creative_ads.size());
```

Comment:
Could we please add a DCHECK inside `ComputeVoteRegistry` to make sure the returned value is of the same size as what is passed in.

Reply:
so remove here and add it there, right?